### DISCUSSION TITLE
Feature Selection

### DISCUSSION THREAD

**Discussion by gm039** [pycaret/pycaret#1785] [Q&A] [answered]
Hi,

I have a query. I would like to know about the feature selected in the k-fold cross-validation using the following command. 

>> RF_Model = create_model('rf',fold = 5) 

Is the features selected for each fold of training data are different? How are the final selected features decided? Are the final selected features different from each fold selected features?

**Comment by gm039** [comment]
@ngupta23 @moezali1 Can you please help me?

**Reply by moezali1** [reply]
Feature selection is done when the `setup` is run. All folds during training get the same `X_train`.

**Comment by ngupta23** [comment]
Each fold is a different RF model (generated using the same hyperparameters but different datasets) so technically they can have a different set of features selected. I am not sure if there is a way to get the model for each fold. You can do return_tuner = True in `tune_model` but it just returns the results as far as I know and not the model for each fold.

@Yard1  Any idea if we can do this?

**Comment by moezali1** [comment, marked as answer]
Feature selection is done when the `setup` is run. All folds during training get the same `X_train`.

**Reply by gm039** [reply]
Thanks, @moezali1 @ngupta23.

I assume the feature selected (when the setup is run) is based on a complete training set (all k fold data). If yes, then using the same selected features may overfit the models trained for each iteration (k-1 folds data) and hence an overfitted cross-validation scores.
As @ngupta23  mentioned, the features may/should be different for each fold and hence they should be selected based upon the k-1 fold data only instead of all training data.

Correct me if I am misinterpreting.

**Comment by ngupta23** [comment]
> As @ngupta23 mentioned, the features may/should be different for each fold and hence they should be selected based upon the k-1 fold data only instead of all training data.

FYI...I was referring to the features selected by Random Forrest in the training process (not referring to preprocessing). This is different for each fold (when cross validation is done), but once CV is done, the model is trained on the entire train data set and returned.
